packages/blue-sdk-viem/src/utils.ts

```diff
@@ -28,3 +44,156 @@ export const safeParseUnits = (strValue: string, decimals = 18) => {
+ *   }),
+ *   // These `args` should be placeholders; just match the type of the actual `args` above
+ *   { abi: blueAbi, name: "idToMarketParams", args: ["0x"] },
+ * )
+ * ```
+ *
+ * @example
+ *
+ * ```
+ * // Use with wagmi hook...
+ * const { data: marketsData } = useReadContracts({
+ *   contracts: marketIds.map(
+ *     (marketId) =>
+ *       ({
+ *         chainId,
+ *         address: morphoAddress,
+ *         abi: morphoAbi,
+ *         functionName: "market",
+ *         args: [marketId],
+ *       }) as const,
+ *   ),
+ *   allowFailure: false,
+ *   query: {
+ *     select(data) {
+ *       // These `args` should be placeholders; just match the type of the actual `args` above
+ *       return data.map((x) => restructure(x, { abi: morphoAbi, name: "market", args: ["0x"] }));
+ *     },
+ *   },
+ * });
+ * ```
+ */
+export function restructure<
+  const abi extends Abi,
+  name extends AbiItemName<abi> & ContractFunctionName<abi, "view" | "pure">,
+  const args extends AbiItemArgs<abi, name>,
+  outputs extends readonly unknown[],
+>(outputs: outputs, parameters: GetAbiItemParameters<abi, name, args>) {
+  const x = getAbiItem(parameters);
+  switch (x?.type) {
+    case "function": {
+      if (x.outputs.some((output) => output.name === undefined)) {
+        throw new Error(
+          `Attempted to restructure return values lacking names in ABI ${parameters.args!} ${x.outputs}`,
+        );
+      }
+      return zipParams(
+        x.outputs as ExtractAbiFunctionForArgs<
+          abi,
+          "view" | "pure",
+          name,
+          args
+        >["outputs"],
+        outputs,
+      );
+    }
+    default:
+      throw new Error(
+        `Attempted to restructure return values for non-function type ${x}`,
+      );
+  }
+}
+
+function isTuple<T>(x: T): x is T extends readonly unknown[] ? T : never {
+  return Array.isArray(x);
+}
+
+/**
+ * When reading contracts, viem converts onchain tuples into arrays -- even when tuple values are named in the ABI.
+ * [They argue](https://viem.sh/docs/faq#why-is-a-contract-function-return-type-returning-an-array-instead-of-an-object)
+ * this information loss is justified, as it eliminates the ambiguity between tuple return types and struct return
+ * types. This wrapper converts viem's arrays back to objects, _as if_ the onchain method returned a struct.
+ *
+ * @see {@link restructure}
+ */
+export async function readContractRestructured<
+  chain extends Chain | undefined,
+  const abi extends Abi,
+  name extends AbiItemName<abi> & ContractFunctionName<abi, "view" | "pure">,
+  const args extends ContractFunctionArgs<abi, "view" | "pure", name>,
+>(
+  client: Client<Transport, chain>,
+  parameters: ReadContractParameters<abi, name, args>,
+) {
+  const outputs = await readContract(client, parameters);
+
+  if (!isTuple(outputs)) {
+    throw new Error(
+      `Attempted to restructure non-tuple return values ${parameters.functionName} -> ${outputs}`,
+    );
```

Comment:
Should we throw actually? Or simply return `outputs`?
But then it complexifies the static types: we need to overload the function type to return outputs when abi item outputs is not an array... Seems do-able but so heavy it's not worth it for now